### PR TITLE
[Quant][PT2E] enable qlinear post op fusion for dynamic quant & qat

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122667

**Description**
Add fusion path for dynamic quant and for QAT.
The following patterns can be matched for static quant with QAT cases:
`qx -> qlinear -> add -> optional relu -> optional type convert -> optional quant`

The following patterns can be matched for dynamic quant cases:
`qx -> qlinear -> add -> optional relu`

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear
python test/inductor/test_cpu_cpp_wrapper.py -k test_qlinear
python test/test_quantization.py -k test_linear_unary
python test/test_quantization.py -k test_linear_binary

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D57655830](https://our.internmc.facebook.com/intern/diff/D57655830)

### CODE REVIEW

File: test/quantization/pt2e/test_x86inductor_quantizer.py [hunk 1]
```diff
@@ -1225,26 +1276,28 @@ def test_linear_binary(self):
...
-                        torch.ops.quantized_decomposed.dequantize_per_tensor.default: 1,
+                        quantize_per_tensor_op: 1,
+                        dequantize_per_tensor_op: 1,
                         # quantize_per_channel for weights are const propagated
                         torch.ops.quantized_decomposed.quantize_per_channel.default: 0,
                         torch.ops.quantized_decomposed.dequantize_per_channel.default: 1,
                     }
                 else:
+                    # convert_pt2e disables duplicate dequant for dynamic quant
+                    num_dequant = 1 if is_dynamic else 2
```
Comment (jgong5) [inline test/quantization/pt2e/test_x86inductor_quantizer.py:1287]: but why things are different for static quant? Can we dedup dequant too?

**Review by jgong5 — APPROVED** [review]

File: test/quantization/pt2e/test_x86inductor_quantizer.py (hunk 1 above)
Comment (Xia-Weiwen) [reply test/quantization/pt2e/test_x86inductor_quantizer.py:1287]: Meta decided to duplicate dq for static quant but not for dynamic quant: https://github.com/pytorch/pytorch/pull/115337#discussion_r1433318916 The discussion is long. The conclusion is not to duplicate dq for dynamic quant

Comment (jgong5) [reply test/quantization/pt2e/test_x86inductor_quantizer.py:1287]: I only saw the discussion on the dynamic quant. Did I miss anything? Do you have the specific link to explain why we dup in the static quant but do not dup in the dynamic quant? I just don't understand why we make things inconsistent here. cc @jerryzh168

Comment (Xia-Weiwen) [reply test/quantization/pt2e/test_x86inductor_quantizer.py:1287]: The discussion was about dynamic quantization only. The difference of behavior between static and dynamic quantization is due to the design XNNPACK workflow. Need comments from @jerryzh168 for more details.

Comment (jerryzh168) [reply test/quantization/pt2e/test_x86inductor_quantizer.py:1390]: there were some discussions here before: https://github.com/pytorch/pytorch/pull/115337#discussion_r1446496919

Comment (jerryzh168) [reply test/quantization/pt2e/test_x86inductor_quantizer.py:1390]: I think dup for static quant should probably be removed as well, this should be done by the lowering pass

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57655830).

**Comment by jerryzh168** [comment]
similar errors:

```
======================================================================
FAIL: test_linear_binary2 (caffe2.test.quantization.pt2e.test_x86inductor_quantizer.TestQuantizePT2EX86Inductor)
Test Pattern:
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/torch/testing/_internal/common_quantization.py", line 404, in wrapper
    fn(*args, **kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/caffe2/test/quantization/pt2e/test_x86inductor_quantizer.py", line 1497, in test_linear_binary2
    fq_m = self._test_quantizer(
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/caffe2/test/quantization/pt2e/test_x86inductor_quantizer.py", line 566, in _test_quantizer
    self.checkGraphModuleNodes(
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/torch/testing/_internal/common_quantization.py", line 811, in checkGraphModuleNodes
    self.assertTrue(
AssertionError: False is not true : Check failed for graph:'get_attr' '_scale_0' '_scale_0' () {}
'get_attr' '_zero_point_0' '_zero_point_0' () {}
'get_attr' '_frozen_param0' '_frozen_param0' () {}
'call_function' 'dequantize_per_channel_default' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')> (_frozen_param0, _scale_0, _zero_point_0, 0, -128, 127, torch.int8) {}
'get_attr' 'linear_bias' 'linear.bias' () {}
'get_attr' '_scale_1' '_scale_1' () {}
'get_attr' '_zero_point_1' '_zero_point_1' () {}
'get_attr' '_frozen_param1' '_frozen_param1' () {}
'call_function' 'dequantize_per_channel_default_1' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')> (_frozen_param1, _scale_1, _zero_point_1, 0, -128, 127, torch.int8) {}
'get_attr' 'linear2_bias' 'linear2.bias' () {}
'placeholder' 'x' 'x' () {}
'call_function' 'quantize_per_tensor_default' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')> (x, 0.014782148413360119, 133, 0, 255, torch.uint8) {}
'call_function' 'dequantize_per_tensor_default' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')> (quantize_per_tensor_default, 0.014782148413360119, 133, 0, 255, torch.uint8) {}
'call_function' 'linear' <OpOverload(op='aten.linear', overload='default')> (dequantize_per_tensor_default, dequantize_per_channel_default, linear_bias) {}
'call_function' 'quantize_per_tensor_default_1' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')> (linear, 0.010831012390553951, 116, 0, 255, torch.uint8) {}
'call_function' 'dequantize_per_tensor_default_1' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')> (quantize_per_tensor_default_1, 0.010831012390553951, 116, 0, 255, torch.uint8) {}
'call_function' 'linear_1' <OpOverload(op='aten.linear', overload='default')> (dequantize_per_tensor_default_1, dequantize_per_channel_default_1, linear2_bias) {}
'call_function' 'add' <OpOverload(op='aten.add', overload='Tensor')> (linear, linear_1) {}
'output' 'output_1' 'output' ((add,),) {}Expected ordered list:['call_function' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')>, 'call_function' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')>, 'call_function' <OpOverload(op='aten.linear', overload='default')>, 'call_function' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')>, 'call_function' <OpOverload(op='aten.add', overload='Tensor')>]

----------------------------------------------------------------------
Ran 1 test in 1.126s

FAILED (failures=1)

AssertionError: False is not true : Check failed for graph:'get_attr' '_scale_0' '_scale_0' () {}
'get_attr' '_zero_point_0' '_zero_point_0' () {}
'get_attr' '_frozen_param0' '_frozen_param0' () {}
'call_function' 'dequantize_per_channel_default' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')> (_frozen_param0, _scale_0, _zero_point_0, 0, -128, 127, torch.int8) {}
'get_attr' 'linear_bias' 'linear.bias' () {}
'get_attr' '_scale_1' '_scale_1' () {}
'get_attr' '_zero_point_1' '_zero_point_1' () {}
'get_attr' '_frozen_param1' '_frozen_param1' () {}
'call_function' 'dequantize_per_channel_default_1' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')> (_frozen_param1, _scale_1, _zero_point_1, 0, -128, 127, torch.int8) {}
'get_attr' 'linear2_bias' 'linear2.bias' () {}
'placeholder' 'x' 'x' () {}
'call_function' 'quantize_per_tensor_default' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')> (x, 0.014782148413360119, 133, 0, 255, torch.uint8) {}
'call_function' 'dequantize_per_tensor_default' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')> (quantize_per_tensor_default, 0.014782148413360119, 133, 0, 255, torch.uint8) {}
'call_function' 'linear' <OpOverload(op='aten.linear', overload='default')> (dequantize_per_tensor_default, dequantize_per_channel_default, linear_bias) {}
'call_function' 'quantize_per_tensor_default_1' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')> (linear, 0.010831012390553951, 116, 0, 255, torch.uint8) {}
'call_function' 'dequantize_per_tensor_default_1' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')> (quantize_per_tensor_default_1, 0.010831012390553951, 116, 0, 255, torch.uint8) {}
'call_function' 'linear_1' <OpOverload(op='aten.linear', overload='default')> (dequantize_per_tensor_default_1, dequantize_per_channel_default_1, linear2_bias) {}
'call_function' 'add' <OpOverload(op='aten.add', overload='Tensor')> (linear, linear_1) {}
'output' 'output_1' 'output' ((add,),) {}Expected ordered list:['call_function' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')>, 'call_function' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')>, 'call_function' <OpOverload(op='aten.linear', overload='default')>, 'call_function' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')>, 'call_function' <OpOverload(op='aten.add', overload='Tensor')>]
  File "/usr/local/fbcode/platform010/lib/python3.10/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/local/fbcode/platform010/lib/python3.10/unittest/case.py", line 591, in run
    self._callTestMethod(testMethod)
  File "/usr/local/fbcode/platform010/lib/python3.10/unittest/case.py", line 549, in _callTestMethod
    method()
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/torch/testing/_internal/common_quantization.py", line 404, in wrapper
    fn(*args, **kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/caffe2/test/quantization/pt2e/test_x86inductor_quantizer.py", line 1497, in test_linear_binary2
    fq_m = self._test_quantizer(
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/caffe2/test/quantization/pt2e/test_x86inductor_quantizer.py", line 566, in _test_quantizer
    self.checkGraphModuleNodes(
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/torch/testing/_internal/common_quantization.py", line 811, in checkGraphModuleNodes
    self.assertTrue(
```


and also for `caffe2/test/quantization:test_quantization - test_linear_binary_unary_serials (caffe2.test.quantization.pt2e.test_x86inductor_quantizer.TestQuantizePT2EX86Inductor)`

**Review by leslie-fang-intel — APPROVED** [review]

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57655830).

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 May I know how it is going with the internal checks? Thanks

**Comment by jerryzh168** [comment]
> similar errors:
> 
> ```
> ======================================================================
> FAIL: test_linear_binary2 (caffe2.test.quantization.pt2e.test_x86inductor_quantizer.TestQuantizePT2EX86Inductor)
> Test Pattern:
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/torch/testing/_internal/common_quantization.py", line 404, in wrapper
>     fn(*args, **kwargs)
>   File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/caffe2/test/quantization/pt2e/test_x86inductor_quantizer.py", line 1497, in test_linear_binary2
>     fq_m = self._test_quantizer(
>   File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/caffe2/test/quantization/pt2e/test_x86inductor_quantizer.py", line 566, in _test_quantizer
>     self.checkGraphModuleNodes(
>   File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/torch/testing/_internal/common_quantization.py", line 811, in checkGraphModuleNodes
>     self.assertTrue(
> AssertionError: False is not true : Check failed for graph:'get_attr' '_scale_0' '_scale_0' () {}
> 'get_attr' '_zero_point_0' '_zero_point_0' () {}
> 'get_attr' '_frozen_param0' '_frozen_param0' () {}
> 'call_function' 'dequantize_per_channel_default' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')> (_frozen_param0, _scale_0, _zero_point_0, 0, -128, 127, torch.int8) {}
> 'get_attr' 'linear_bias' 'linear.bias' () {}
> 'get_attr' '_scale_1' '_scale_1' () {}
> 'get_attr' '_zero_point_1' '_zero_point_1' () {}
> 'get_attr' '_frozen_param1' '_frozen_param1' () {}
> 'call_function' 'dequantize_per_channel_default_1' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')> (_frozen_param1, _scale_1, _zero_point_1, 0, -128, 127, torch.int8) {}
> 'get_attr' 'linear2_bias' 'linear2.bias' () {}
> 'placeholder' 'x' 'x' () {}
> 'call_function' 'quantize_per_tensor_default' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')> (x, 0.014782148413360119, 133, 0, 255, torch.uint8) {}
> 'call_function' 'dequantize_per_tensor_default' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')> (quantize_per_tensor_default, 0.014782148413360119, 133, 0, 255, torch.uint8) {}
> 'call_function' 'linear' <OpOverload(op='aten.linear', overload='default')> (dequantize_per_tensor_default, dequantize_per_channel_default, linear_bias) {}
> 'call_function' 'quantize_per_tensor_default_1' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')> (linear, 0.010831012390553951, 116, 0, 255, torch.uint8) {}
> 'call_function' 'dequantize_per_tensor_default_1' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')> (quantize_per_tensor_default_1, 0.010831012390553951, 116, 0, 255, torch.uint8) {}
> 'call_function' 'linear_1' <OpOverload(op='aten.linear', overload='default')> (dequantize_per_tensor_default_1, dequantize_per_channel_default_1, linear2_bias) {}
> 'call_function' 'add' <OpOverload(op='aten.add', overload='Tensor')> (linear, linear_1) {}
> 'output' 'output_1' 'output' ((add,),) {}Expected ordered list:['call_function' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')>, 'call_function' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')>, 'call_function' <OpOverload(op='aten.linear', overload='default')>, 'call_function' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')>, 'call_function' <OpOverload(op='aten.add', overload='Tensor')>]
> 
> ----------------------------------------------------------------------
> Ran 1 test in 1.126s
> 
> FAILED (failures=1)
> 
> AssertionError: False is not true : Check failed for graph:'get_attr' '_scale_0' '_scale_0' () {}
> 'get_attr' '_zero_point_0' '_zero_point_0' () {}
> 'get_attr' '_frozen_param0' '_frozen_param0' () {}
> 'call_function' 'dequantize_per_channel_default' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')> (_frozen_param0, _scale_0, _zero_point_0, 0, -128, 127, torch.int8) {}
> 'get_attr' 'linear_bias' 'linear.bias' () {}
> 'get_attr' '_scale_1' '_scale_1' () {}
> 'get_attr' '_zero_point_1' '_zero_point_1' () {}
> 'get_attr' '_frozen_param1' '_frozen_param1' () {}
> 'call_function' 'dequantize_per_channel_default_1' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')> (_frozen_param1, _scale_1, _zero_point_1, 0, -128, 127, torch.int8) {}
> 'get_attr' 'linear2_bias' 'linear2.bias' () {}
> 'placeholder' 'x' 'x' () {}
> 'call_function' 'quantize_per_tensor_default' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')> (x, 0.014782148413360119, 133, 0, 255, torch.uint8) {}
> 'call_function' 'dequantize_per_tensor_default' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')> (quantize_per_tensor_default, 0.014782148413360119, 133, 0, 255, torch.uint8) {}
> 'call_function' 'linear' <OpOverload(op='aten.linear', overload='default')> (dequantize_per_tensor_default, dequantize_per_channel_default, linear_bias) {}
> 'call_function' 'quantize_per_tensor_default_1' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')> (linear, 0.010831012390553951, 116, 0, 255, torch.uint8) {}
> 'call_function' 'dequantize_per_tensor_default_1' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')> (quantize_per_tensor_default_1, 0.010831012390553951, 116, 0, 255, torch.uint8) {}
> 'call_function' 'linear_1' <OpOverload(op='aten.linear', overload='default')> (dequantize_per_tensor_default_1, dequantize_per_channel_default_1, linear2_bias) {}
> 'call_function' 'add' <OpOverload(op='aten.add', overload='Tensor')> (linear, linear_1) {}
> 'output' 'output_1' 'output' ((add,),) {}Expected ordered list:['call_function' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')>, 'call_function' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')>, 'call_function' <OpOverload(op='aten.linear', overload='default')>, 'call_function' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')>, 'call_function' <OpOverload(op='aten.add', overload='Tensor')>]
>   File "/usr/local/fbcode/platform010/lib/python3.10/unittest/case.py", line 59, in testPartExecutor
>     yield
>   File "/usr/local/fbcode/platform010/lib/python3.10/unittest/case.py", line 591, in run
>     self._callTestMethod(testMethod)
>   File "/usr/local/fbcode/platform010/lib/python3.10/unittest/case.py", line 549, in _callTestMethod
>     method()
>   File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/torch/testing/_internal/common_quantization.py", line 404, in wrapper
>     fn(*args, **kwargs)
>   File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/caffe2/test/quantization/pt2e/test_x86inductor_quantizer.py", line 1497, in test_linear_binary2
>     fq_m = self._test_quantizer(
>   File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/caffe2/test/quantization/pt2e/test_x86inductor_quantizer.py", line 566, in _test_quantizer
>     self.checkGraphModuleNodes(
>   File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5839ccd20a633894/caffe2/test/quantization/__test_quantization__/test_quantization#link-tree/torch/testing/_internal/common_quantization.py", line 811, in checkGraphModuleNodes
>     self.assertTrue(
> ```
> 
> and also for `caffe2/test/quantization:test_quantization - test_linear_binary_unary_serials (caffe2.test.quantization.pt2e.test_x86inductor_quantizer.TestQuantizePT2EX86Inductor)`

@Xia-Weiwen these two failures are still there I think

**Comment by jerryzh168** [comment]
caffe2/test/quantization:test_quantization - test_linear_binary_unary_serials (caffe2.test.quantization.pt2e.test_x86inductor_quantizer.TestQuantizePT2EX86Inductor)
caffe2/test/quantization:test_quantization - test_linear_binary2 (caffe2.test.quantization.pt2e.test_x86inductor_quantizer.TestQuantizePT2EX86Inductor)

full error: https://gist.github.com/jerryzh168/cd57ab46aec8930ca458654482205ca7

**Comment by Xia-Weiwen** [comment]
@jerryzh168 Thanks a lot. I will check again. BTW, do you have any idea about differences of exported graphs between internal and public cases?

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 The error was raised when checking `expected ordered list` in the model graph. However, I saw the `expected ordered list` in the error log is different from what is in the latest code, but rather the same as that before I made changes to the code last time. I have rebased this PR without new changes. Could you please import it again and double check if the imported patch is correct? Thanks.

---
For example, for the case `test_linear_binary2`, the `expected ordered list` in the error log is (note that **dequantize_per_channel is in the *fourth* place**)
```
Expected ordered list:[
'call_function' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')>,
'call_function' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')>,
'call_function' <OpOverload(op='aten.linear', overload='default')>,
'call_function' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')>,
'call_function' <OpOverload(op='aten.add', overload='Tensor')>]
```
However, the list in the code is (note that **dequantize_per_channel is in the *first* place**):
```python
node_list = [
    torch.ops.quantized_decomposed.dequantize_per_channel.default,
    quantize_per_tensor_op,  # quantize_per_tensor.default or .tensor
    dequantize_per_tensor_op,  # dequantize_per_tensor.default or .tensor
    torch.ops.aten.linear.default,
    torch.ops.aten.add_.Tensor
    if inplace_add
    else torch.ops.aten.add.Tensor,  # inplace_add is always false
]
```

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57655830).

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 May I know the result of your last import? Thanks!

**Comment by jerryzh168** [comment]
@Xia-Weiwen need to rebase and resolve conflict

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57655830).

**Comment by Xia-Weiwen** [comment]
@jerryzh168 I have been off recently. Thanks for merging this PR!

**Pull request closed without merging**